pkg/netstore/netstore.go

```diff
@@ -58,6 +60,17 @@ func New(s storage.Storer, validStamp postage.ValidStampFn, rcb recovery.Callbac
 // local-store.
 func (s *store) Get(ctx context.Context, mode storage.ModeGet, addr swarm.Address) (ch swarm.Chunk, err error) {
 	ch, err = s.Storer.Get(ctx, mode, addr)
+	if err == nil {
+		// ensure the chunk we get locally is valid. If not, retrieve the chunk
+		// from network. If there is any corruption of data in the local storage,
+		// this would ensure it is retrieved again from network and added back with
+		// the correct data
+		if !cac.Valid(ch) && !soc.Valid(ch) {
+			err = errors.New("invalid chunk")
```

Comment:
problem is, this error will not let this go to retrieval (see below L75-91) but wraps it as netstore error and returns (L92)

Reply:
Sorry for that. Thanks for catching this!

The current master is running with this fix and passing all iterations. So ideally we don't need this additional check. But we all think this is currently nice to have as this will ensure download doesnt return garbage. The hit in performance is not so much.

Reply:
I have added a test to verify this as well.

Reply:
In fact I would like to not have this check at all. We trust our own DB and chunker. 
The circumstance that we introduced a bug does not justify a check in the production code.

Reply:
@zelig    cc @metacertain @janos 
The problem is this issue was very difficult to identify by going through the code. There are multiple things that led to this together. Secondly there are other reasons why a disk could potentially get corrupted without us introducing any code changes as well.

Given all this, we thought it would be okay to not trust local storage as well. This was thought to be in-line with other assumptions that we make. For eg. we have this check in the protocols already. So during download this would be happening either way. In the interest of doing the release at the earliest, would it be okay to keep this change now and maybe have a discussion internally whether or not to keep it in the upcoming release?

We are already ready with the changes for the next version. This can be done immediately.